src/main/scala/devices/debug/Debug.scala

```diff
@@ -289,11 +297,110 @@ class TLDebugModuleOuter(device: Device)(implicit p: Parameters) extends LazyMod
+      val HAWINDOWSELReg = Wire(init = new HAWINDOWSELFields().fromBits(AsyncResetReg(updateData = HAWINDOWSELNxt.asUInt,
+        resetData = 0,
+        enable = true.B,
+        name = "HAWINDOWSEL"
+      )))
+
+      HAWINDOWSELRdData := Wire(init = HAWINDOWSELReg)
+      HAWINDOWSELNxt := HAWINDOWSELReg
+      when (~dmactive | ~supportHartArray.B) {
+        HAWINDOWSELNxt := HAWINDOWSELReset
+      } .otherwise {
+        when (HAWINDOWSELWrEn) {
+            // Unneeded upper bits of HAWINDOWSEL are tied to 0.  Entire register is 0 if 32 or fewer harts.
+          if (nComponents > 32) {
+            HAWINDOWSELNxt.hawindowsel := HAWINDOWSELWrData.hawindowsel & (log2Up(nComponents) - 5).U
+          } else {
+            HAWINDOWSELNxt.hawindowsel := 0.U
+          }
+        }
+      }
+
+      val numHAMASKSlices = ((nComponents - 1)/32)+1
+      HAWINDOWRdData.maskdata := 0.U     // default, overridden below
+      for (ii <- 0 until numHAMASKSlices) {
+        val SliceMask = if (nComponents > ii*32 + 31) 0xFFFFFFFF  // All harts in this slice exist
+                        else (1<<(nComponents - ii*32)) - 1       // Partial last slice
+        val HAMASKRst = Wire(init = (new HAWINDOWFields().fromBits(0.U)))
+        val HAMASKNxt = Wire(init = (new HAWINDOWFields().fromBits(0.U)))
+        val HAMASKReg = Wire(init = Vec(AsyncResetReg(updateData = HAMASKNxt.asUInt,
+          resetData = 0,
+          enable = true.B,
+          name = "HAMASK")))
+
+        when (ii.U === HAWINDOWSELReg.hawindowsel) {
+          HAWINDOWRdData.maskdata := HAMASKReg.asUInt & SliceMask.U
+        }
+
+        HAMASKNxt.maskdata := HAMASKReg.asUInt
+        when (~dmactive) {
+          HAMASKNxt := HAMASKRst
+        }.otherwise {
+          when (HAWINDOWWrEn && (ii.U === HAWINDOWSELReg.hawindowsel)) {
+            HAMASKNxt.maskdata := HAWINDOWWrData.maskdata
+          }
+        }
+
+         // drive each slice of hamask with HAMASKReg or with new value being written
+        for (jj <- 0 until 32) {
+          if (ii*32 + jj < nComponents) {
+            hamask(ii*32 + jj) := Mux(HAWINDOWWrEn && (ii.U === HAWINDOWSELReg.hawindowsel),
+                                      (HAWINDOWWrData.maskdata >> jj) & 1.U,
+                                      (HAMASKReg.asUInt >> jj) & 1.U)
+          }
+        }
+      }
+    }
+
+    //--------------------------------------------------------------
+    // DMI register decoder for Outer
+    //--------------------------------------------------------------
+      // regmap addresses are byte offsets from lowest address
+    def DMI_DMCONTROL_OFFSET   = 0
+    def DMI_HAWINDOWSEL_OFFSET = ((DMI_HAWINDOWSEL - DMI_DMCONTROL) << 2)
+    def DMI_HAWINDOW_OFFSET    = ((DMI_HAWINDOW - DMI_DMCONTROL) << 2)
+
+    if (supportHartArray) {
+      dmiNode.regmap(
+        DMI_DMCONTROL_OFFSET   -> Seq(RWNotify(32, DMCONTROLRdData.asUInt(),
```

Comment:
can you build up this sequence separately, and conditionally add to it, so that you don't have to copy the code for DMI_DMCONTROL_OFFSET. Then you can just call `dmiNode.regmap(prebuild_sequence)`

Reply:
I could use some help with the code for this suggestion...

Reply:
Maybe simpler, you can always have the offset but put the Seq of RegField there or not:

`DMI_HAWINDOWSEL_OFFSET -> (if (supportsHartArray) Seq(...) else Nil)`